File: server/src/main/java/edu/cnm/deepdive/coffeeshop/service/FavoriteService.java
package edu.cnm.deepdive.coffeeshop.service;

import edu.cnm.deepdive.coffeeshop.model.dto.FavoriteRequestDto;
import edu.cnm.deepdive.coffeeshop.model.dto.ShopDto;
import edu.cnm.deepdive.coffeeshop.model.entity.Profile;
import edu.cnm.deepdive.coffeeshop.model.entity.Shop;
import java.util.List;
import java.util.Set;
import java.util.UUID;

public interface FavoriteService {

  List<ShopDto> getFavorites(Profile profile);
  // TODO: 7/30/26 Verify returning shops is okay.

  ShopDto saveFavorite(FavoriteRequestDto dto, Profile profile);
  // TODO: 7/30/26 Verify that returning nothing is okay.

  void removeFavorite(UUID  shopId, Profile profile);
  // TODO: 7/30/26 Verify that returning nothing is okay.
}

File: app/src/main/java/edu/cnm/deepdive/coffeeshop/service/FavoriteService.kt
package edu.cnm.deepdive.coffeeshop.service

import edu.cnm.deepdive.coffeeshop.model.domain.Shop
import edu.cnm.deepdive.coffeeshop.repository.FavoriteRepository
import jakarta.inject.Inject
import jakarta.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.future.future
import java.util.concurrent.CompletableFuture

@Singleton
class FavoriteService @Inject constructor(private val favoriteRepository: FavoriteRepository){

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    suspend fun addFavorite(shop: Shop)=
        scope.future {
            favoriteRepository.addFavorite(shop)
        }

    suspend fun removeFavorite(shop: Shop)=
        scope.future {
            favoriteRepository.removeFavorite(shop)
        }

    suspend fun getFavorites(): CompletableFuture<List<Shop>> =
        scope.future {
        favoriteRepository.getFavorites()
    }
}